File: ArduinoWiFly/NetduinoWiFly/ParsedStream.cs
namespace NetduinoWiFly
{
	using System;

	internal class ParsedStream
	{
		#region Constants
		public const int RX_BUFFER_SIZE = 64;
		public const string MATCH_TOKEN = "*CLOS*";
		#endregion
		#region Fields
		private byte _bytesMatched;
		private bool _closed;
		private RingBuffer _rxBuffer;
		private SpiUartDevice _uart;
		#endregion
		#region Properties
		private int FreeSpace
		{
			get { return RX_BUFFER_SIZE - Available(true) - 1; }
		}
		public bool Closed { get { return _closed && Available() != 0; } }
		#endregion

		#region Constructors
		public ParsedStream(SpiUartDevice uartDevice)
		{
			_uart = uartDevice;
			Reset();
		}
		#endregion

		#region Methods
		private byte Available(bool raw)
		{
			byte availablBytes = (byte)((RX_BUFFER_SIZE + _rxBuffer.Head - _rxBuffer.Tail) % RX_BUFFER_SIZE);

			if(!raw)
			{
				if (availablBytes > _bytesMatched)
					availablBytes -= _bytesMatched;
				else
					availablBytes = 0;
			}

			return availablBytes;
		}

		public byte Available()
		{
			while (!_closed && FreeSpace != 0 && _uart.AvailableBytes != 0)
			{
				GetByte();
			}

			return Available(false);
		}

		private void GetByte()
		{
			int i;

			if (_closed || FreeSpace == 0)
				return;

			i = _uart.Read();
			if (i == -1)
				return;

			if (i == MATCH_TOKEN[_bytesMatched])
			{
				_bytesMatched++;
				if (_bytesMatched == MATCH_TOKEN.Length)
					_closed = true;
			}
			else if (i == MATCH_TOKEN[0])
				_bytesMatched = 1;
			else
				_bytesMatched = 0;

			StoreByte(i);
		}

		public int Read()
		{
			if (Available() != 0)
				GetByte();

			if (Available() != 0)
				return -1;

			byte b = _rxBuffer.Buffer[_rxBuffer.Tail];
			_rxBuffer.Tail = (_rxBuffer.Tail + 1) % RX_BUFFER_SIZE;
			return b;
		}

		public void Reset()
		{
			_rxBuffer = new RingBuffer();
			_closed = false;
			_bytesMatched = 0;
		}

		private void StoreByte(int i)
		{
			if (i < 0 || i > byte.MaxValue)
				throw new ArgumentException("Must be within the range of byte.", "i");
			StoreByte((byte)i);
		}

		private void StoreByte(byte b)
		{
			int i = (_rxBuffer.Head + 1) % RX_BUFFER_SIZE;

			if(i != _rxBuffer.Tail)
			{
				_rxBuffer.Buffer[_rxBuffer.Head] = b;
				_rxBuffer.Head = i;
			}
		}
		#endregion


		#region Nested Types
		private class RingBuffer
		{
			#region Fields
			public byte[] Buffer { get; set; }
			public int Head { get; set; }
			public int Tail { get; set; }
			#endregion

			#region Constructors
			public RingBuffer()
			{
				Buffer = new byte[RX_BUFFER_SIZE];
			}
			#endregion
		}
		#endregion
	}
}

File: ArduinoWiFly/ArduinoWiFly/ParsedStream.cpp
#include "ParsedStream.h"

void ParsedStream::storeByte(unsigned char c) 
{
    int i = (_rx_buffer.head + 1) % RX_BUFFER_SIZE;

    if (i != _rx_buffer.tail) 
	{
        _rx_buffer.buffer[_rx_buffer.head] = c;
        _rx_buffer.head = i;
    }
}

ParsedStream::ParsedStream(SpiUartDevice& uart) : _uart(uart) { reset(); }

void ParsedStream::reset() 
{
    ring_buffer _rx_buffer = { { 0 }, 0, 0};
    _closed = false;
    bytes_matched = 0;    
}

uint8_t ParsedStream::available(bool raw) 
{
    uint8_t available_bytes;
    
    available_bytes = (RX_BUFFER_SIZE + _rx_buffer.head - _rx_buffer.tail) % RX_BUFFER_SIZE;

    if (!raw) 
	{
        if (available_bytes > bytes_matched) 
            available_bytes -= bytes_matched;
        else 
            available_bytes = 0;
    }

    return available_bytes;
}

uint8_t ParsedStream::available() 
{
    while (!_closed && freeSpace() && _uart.available())
        getByte();

    return available(false);
}

bool ParsedStream::closed() 
{
    return _closed && !available();
}

int ParsedStream::read(void) 
{
    if (!available())
        getByte();

    if (!available())
        return -1;
	else 
	{
        unsigned char c = _rx_buffer.buffer[_rx_buffer.tail];
        _rx_buffer.tail = (_rx_buffer.tail + 1) % RX_BUFFER_SIZE;
        return c;
    }
}

int ParsedStream::freeSpace() 
{
    return RX_BUFFER_SIZE - available(true) - 1;
}

void ParsedStream::getByte() 
{
    int c;

    if (_closed)
        return;

    if (freeSpace() == 0)
        return;

    c = _uart.read();

    if (c == -1)
        return;

    if (c == MATCH_TOKEN[bytes_matched])
	{
        bytes_matched++;
        if (bytes_matched == strlen(MATCH_TOKEN))
            _closed = true;
    } 
	else if (c == MATCH_TOKEN[0])
        bytes_matched = 1;
    else
        bytes_matched = 0;

    storeByte(c);    
}